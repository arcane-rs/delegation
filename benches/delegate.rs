use criterion::{criterion_group, criterion_main, Criterion};
use delegation::delegate;

#[delegate]
trait AsStr {
    fn as_str(&self) -> &str;
}

impl AsStr for String {
    fn as_str(&self) -> &str {
        self
    }
}

macro_rules! def_enum {
    ($($variant:ident),+) => {
        #[delegate(derive(AsStr))]
        enum Name {
            $($variant(String)),+
        }
    };
}

macro_rules! def_bench {
    ($($variant:ident),*) => {
        fn bench(bencher: &mut ::criterion::Bencher) {
            $(
                bencher.iter(|| {
                    let name = ::criterion::black_box(
                        Name::$variant(String::from("Test"))
                    );

                    assert_eq!(name.as_str(), "Test");
                });
            )*
        }
    };
}

macro_rules! define {
    ($($variant:ident),*) => {
        def_enum!($($variant),*);
        def_bench!($($variant),*);
    };
}

define!(
    A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y,
    Z
);

fn delegate_benchmark(c: &mut Criterion) {
    c.bench_function("delegate", bench);
}

criterion_group!(benches, delegate_benchmark);
criterion_main!(benches);
