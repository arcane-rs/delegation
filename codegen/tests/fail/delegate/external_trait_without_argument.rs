use delegation_codegen::delegate;

#[delegate]
pub trait AsRefDef<T: ?Sized> {
    fn as_ref(&self) -> &T;
}

#[delegate(derive(for<> AsRef<str> as AsRefDef))]
pub enum Name {
    First(String),
}

fn main() {
    unreachable!()
}
