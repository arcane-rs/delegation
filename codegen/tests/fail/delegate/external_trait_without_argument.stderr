error[E0576]: cannot find associated type `Wrapper` in trait `AsRefDef`
 --> tests/fail/delegate/external_trait_without_argument.rs:3:1
  |
3 | #[delegate]
  | ^^^^^^^^^^^ not found in `AsRefDef`
...
8 | #[delegate(derive(AsRef<str> as AsRefDef))]
  | ------------------------------------------- in this procedural macro expansion
  |
  = note: this error originates in the macro `::delegation::private::impl_for` which comes from the expansion of the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0220]: associated type `__delegate_AsRefDef__Bind0` not found for `Self`
 --> tests/fail/delegate/external_trait_without_argument.rs:4:11
  |
4 | pub trait AsRefDef<T: ?Sized> {
  |           ^^^^^^^^ there is an associated type `__delegate_AsRefDef__Bind0` in the trait `__delegate_AsRefDef__Scope`
...
8 | #[delegate(derive(AsRef<str> as AsRefDef))]
  | ------------------------------------------- in this procedural macro expansion
  |
  = note: this error originates in the macro `AsRefDef` which comes from the expansion of the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
